lib/elixir/lib/map.ex

```diff
@@ -93,11 +95,18 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.to_list
```

Comment:
I should note that because of https://github.com/elixir-lang/elixir/blob/master/lib/elixir/lib/enum.ex#L2269 I have been able to avoid the double reversal issue in my benchmarking so far. I will add some tests that use ranges as well and see what can be done.